File: createPdfTrackingList.js
const fs = require('fs').promises;
const path = require('path');

async function trackConfPdfs() {
  try {
    console.log('Starting PDF tracking process...');
    
    // Read the conference API results file
    const filePath = path.join(__dirname, 'assembly_bills_conference_api_results.json');
    const fileContent = await fs.readFile(filePath, 'utf8');
    const data = JSON.parse(fileContent);
    
    if (!data.results || !Array.isArray(data.results)) {
      throw new Error('No results array found in the file');
    }
    
    console.log(`Found ${data.results.length} total results`);
    
    // Filter results that have api_response with data
    const validResults = data.results.filter(result => 
      result.api_response && 
      Array.isArray(result.api_response) && 
      result.api_response.length >= 2 &&
      result.api_response[1] &&
      result.api_response[1].row &&
      Array.isArray(result.api_response[1].row)
    );
    
    console.log(`Found ${validResults.length} results with valid API responses`);
    
    // Create downloads directory path (for reference)
    const downloadsDir = path.join(__dirname, 'pdf_downloads');
    
    let totalTracked = 0;
    let totalSkipped = 0;
    const pdfTrackingData = [];
    
    for (const result of validResults) {
      try {
        const rows = result.api_response[1].row;
        
        if (rows.length === 0) {
          console.log(`No conference data found for this result`);
          totalSkipped++;
          continue;
        }
        
        // Get BILL_ID from the first row (all rows should have the same BILL_ID)
        const billId = rows[0].BILL_ID;
        
        if (!billId) {
          console.log('No BILL_ID found in response');
          totalSkipped++;
          continue;
        }
        
        console.log(`Processing BILL_ID: ${billId}`);
        
        // Define directory for this BILL_ID
        const billDir = path.join(downloadsDir, billId);
        
        // Track PDFs for each row
        for (let i = 0; i < rows.length; i++) {
          const row = rows[i];
          
          if (!row.DOWN_URL) {
            console.log(`  No download URL for row ${i + 1}`);
            continue;
          }
          
          try {
            // Create filename from all conference data fields
            const confKnd = (row.CONF_KND || '_').replace(/[<>:"/\\|?*]/g, '_');
            const confId = (row.CONF_ID || '_').replace(/[<>:"/\\|?*]/g, '_');
            const eraco = (row.ERACO || '_').replace(/[<>:"/\\|?*]/g, '_');
            const sess = (row.SESS || '_').replace(/[<>:"/\\|?*]/g, '_');
            const dgr = (row.DGR || '_').replace(/[<>:"/\\|?*]/g, '_');
            const confDt = (row.CONF_DT || '_').replace(/[<>:"/\\|?*]/g, '_').trim();
            
            const filename = `${confKnd}_${confId}_${eraco}_${sess}_${dgr}_${confDt}.pdf`;
            const fullFilePath = path.join(billDir, filename);
            
            // Check if file already exists
            let fileExists = false;
            try {
              await fs.access(fullFilePath);
              fileExists = true;
              console.log(`  ✓ File already exists: ${filename}`);
            } catch (error) {
              // File doesn't exist
              console.log(`  ○ File to be downloaded: ${filename}`);
            }
            
            // Track this PDF
            pdfTrackingData.push({
              bill_id: billId,
              bill_name: row.BILL_NM,
              conference_kind: row.CONF_KND,
              conference_id: row.CONF_ID,
              eraco: row.ERACO,
              session: row.SESS,
              degree: row.DGR,
              conference_date: row.CONF_DT,
              download_url: row.DOWN_URL,
              filename: filename,
              full_path: fullFilePath,
              relative_path: path.join('pdf_downloads', billId, filename),
              file_exists: fileExists,
              tracked_date: new Date().toISOString()
            });
            
            totalTracked++;
            
          } catch (trackingError) {
            console.error(`  ✗ Error tracking ${row.DOWN_URL}:`, trackingError.message);
          }
        }
        
      } catch (resultError) {
        console.error(`Error processing result:`, resultError.message);
      }
    }
    
    // Save tracking data to JSON file
    const trackingData = {
      summary: {
        total_pdfs_tracked: totalTracked,
        total_existing_files: pdfTrackingData.filter(p => p.file_exists).length,
        total_to_download: pdfTrackingData.filter(p => !p.file_exists).length,
        total_skipped: totalSkipped,
        generated_date: new Date().toISOString()
      },
      pdfs: pdfTrackingData
    };
    
    const trackingFilePath = path.join(__dirname, 'pdf_tracking_list.json');
    await fs.writeFile(trackingFilePath, JSON.stringify(trackingData, null, 2), 'utf8');
    
    console.log('\n=== PDF Tracking Summary ===');
    console.log(`Total PDFs tracked: ${totalTracked}`);
    console.log(`Files already exist: ${trackingData.summary.total_existing_files}`);
    console.log(`Files to download: ${trackingData.summary.total_to_download}`);
    console.log(`Total skipped: ${totalSkipped}`);
    console.log(`Tracking data saved to: ${trackingFilePath}`);
    
  } catch (error) {
    console.error('Error in trackConfPdfs:', error);
    throw error;
  }
}

// Run the script if this file is executed directly
if (require.main === module) {
  trackConfPdfs()
    .then(() => {
      console.log('PDF tracking process completed successfully!');
    })
    .catch(error => {
      console.error('PDF tracking process failed:', error);
      process.exit(1);
    });
}

module.exports = { trackConfPdfs };

File: createPdfTrackingList.py
import json
import re
from datetime import datetime
from pathlib import Path
from typing import List, Dict, Any

class PdfTracker:
    def __init__(self):
        self.base_dir = Path(__file__).parent
        self.downloads_dir = self.base_dir / 'pdf_downloads'

    def sanitize_filename(self, text: str) -> str:
        """Sanitize text for use in filename"""
        if not text:
            return '_'
        # Remove invalid filename characters
        sanitized = re.sub(r'[<>:"/\\|?*]', '_', str(text))
        return sanitized.strip() or '_'

    def track_conference_pdfs(self) -> None:
        """Create PDF tracking list without downloading"""
        try:
            print('Starting PDF tracking process...')
            
            # Read the conference API results file
            file_path = self.base_dir / 'assembly_bills_conference_api_results.json'
            with open(file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
            
            if 'results' not in data or not isinstance(data['results'], list):
                raise Exception('No results array found in the file')
            
            print(f'Found {len(data["results"])} total results')
            
            # Filter results that have api_response with data
            valid_results = []
            for result in data['results']:
                try:
                    if (result.get('api_response') and 
                        isinstance(result['api_response'], list) and 
                        len(result['api_response']) >= 2 and
                        result['api_response'][1] and
                        'row' in result['api_response'][1] and
                        isinstance(result['api_response'][1]['row'], list)):
                        valid_results.append(result)
                except:
                    continue
            
            print(f'Found {len(valid_results)} results with valid API responses')
            
            total_tracked = 0
            total_skipped = 0
            pdf_tracking_data = []
            
            for result in valid_results:
                try:
                    rows = result['api_response'][1]['row']
                    
                    if not rows:
                        print('No conference data found for this result')
                        total_skipped += 1
                        continue
                    
                    # Get BILL_ID from the first row (all rows should have the same BILL_ID)
                    bill_id = rows[0].get('BILL_ID')
                    
                    if not bill_id:
                        print('No BILL_ID found in response')
                        total_skipped += 1
                        continue
                    
                    print(f'Processing BILL_ID: {bill_id}')
                    
                    # Define directory for this BILL_ID
                    bill_dir = self.downloads_dir / bill_id
                    
                    # Track PDFs for each row
                    for i, row in enumerate(rows):
                        if not row.get('DOWN_URL'):
                            print(f'  No download URL for row {i + 1}')
                            continue
                        
                        try:
                            # Create filename from all conference data fields
                            conf_knd = self.sanitize_filename(row.get('CONF_KND', '_'))
                            conf_id = self.sanitize_filename(row.get('CONF_ID', '_'))
                            eraco = self.sanitize_filename(row.get('ERACO', '_'))
                            sess = self.sanitize_filename(row.get('SESS', '_'))
                            dgr = self.sanitize_filename(row.get('DGR', '_'))
                            conf_dt = self.sanitize_filename(row.get('CONF_DT', '_'))
                            
                            filename = f'{conf_knd}_{conf_id}_{eraco}_{sess}_{dgr}_{conf_dt}.pdf'
                            full_file_path = bill_dir / filename
                            
                            # Check if file already exists
                            file_exists = full_file_path.exists()
                            if file_exists:
                                print(f'  ✓ File already exists: {filename}')
                            else:
                                print(f'  ○ File to be downloaded: {filename}')
                            
                            # Track this PDF
                            pdf_tracking_data.append({
                                'bill_id': bill_id,
                                'bill_name': row.get('BILL_NM'),
                                'conference_kind': row.get('CONF_KND'),
                                'conference_id': row.get('CONF_ID'),
                                'eraco': row.get('ERACO'),
                                'session': row.get('SESS'),
                                'degree': row.get('DGR'),
                                'conference_date': row.get('CONF_DT'),
                                'download_url': row['DOWN_URL'],
                                'filename': filename,
                                'full_path': str(full_file_path),
                                'relative_path': str(Path('pdf_downloads') / bill_id / filename),
                                'file_exists': file_exists,
                                'tracked_date': datetime.now().isoformat()
                            })
                            
                            total_tracked += 1
                            
                        except Exception as tracking_error:
                            print(f'  ✗ Error tracking {row["DOWN_URL"]}: {tracking_error}')
                    
                except Exception as result_error:
                    print(f'Error processing result: {result_error}')
            
            # Save tracking data to JSON file
            tracking_data = {
                'summary': {
                    'total_pdfs_tracked': total_tracked,
                    'total_existing_files': sum(1 for p in pdf_tracking_data if p['file_exists']),
                    'total_to_download': sum(1 for p in pdf_tracking_data if not p['file_exists']),
                    'total_skipped': total_skipped,
                    'generated_date': datetime.now().isoformat()
                },
                'pdfs': pdf_tracking_data
            }
            
            tracking_file_path = self.base_dir / 'pdf_tracking_list.json'
            with open(tracking_file_path, 'w', encoding='utf-8') as f:
                json.dump(tracking_data, f, ensure_ascii=False, indent=2)
            
            print('\n=== PDF Tracking Summary ===')
            print(f'Total PDFs tracked: {total_tracked}')
            print(f'Files already exist: {tracking_data["summary"]["total_existing_files"]}')
            print(f'Files to download: {tracking_data["summary"]["total_to_download"]}')
            print(f'Total skipped: {total_skipped}')
            print(f'Tracking data saved to: {tracking_file_path}')
            
        except Exception as error:
            print(f'Error in track_conference_pdfs: {error}')
            raise error

def main():
    """Main function to run the PDF tracker"""
    tracker = PdfTracker()
    tracker.track_conference_pdfs()
    print('PDF tracking process completed successfully!')

if __name__ == "__main__":
    main()